Reviving load

blue442 opened the pull request:
Addressing #379, implemented changes to separate the loading of the metadata and dataset from the instantiation of a foundry object in order to facilitate browsing/searching of datasets _prior_ to downloading. In addition, I refactored the downloading of metadata and data to utilize clearer verbiage when calling the functions.

## Specific changes
- Made `load()` a public function again
- Removed the `metadata` arg from `load()`
- Removed `if metadata: res = metadata` from `load()` method
- if `is_doi(name) and not metadata:` -> `if is_doi(name):`
- Remove call to `load()` from the `__init__()` function
- Changed `f.load()` -> `f.fetch_data()`
- Changed `download=True` to `metadata_only=False`

## New usage example snippet

```
import foundry
f = foundry.Foundry()
f.search()
f.fetch_data('foundry_assorted_computational_band_gaps_v1.1', metadata_only=True)
f.download_full_dataset()
data_dictionary = f.load_data()
```

## Documentation notes:
- `load_data()` loads the data into memory, while `download_data()` downloads it to disk, correct? If so, the docstring should probably reflect this.
- Moved 'todos' from code into docstrings (google-style)

## TODO:
- Need to update gitbook
- Need to update notebooks
- From `__init__()` - remove `metadata (dict): **For debug purposes.** A search result analog to prepopulate metadata.`
- Not sure about "# TODO: Creating a new Foundry instance is a problematic way to update the metadata, we should find a way to abstract this." (I don't understand this comment)
- Why do we only return the first of multiple search results? We're already getting all of the results and just throwing all but the first away - should we handle this differently?
- Calling `load()` (now `fetch_data()`)- does a user need to pass in the 'source_id' generated by `list()` as the name argument, or can they pass in the entry under 'name'? ('name' doesn't seem to work)

ascourtas reviewed (changes requested):
Hi Steve! I was discussing this solution with @blaiszik, and he pointed out that there wouldn't be a case where the user would want to download the dataset but not load the metadata (but there are cases where a user _would_ want to load the metadata, but not download the dataset). He suggested the following refactor which I think is very smooth:

> "I think we should just make the metadata loading occur without user intervention. It’s a very cheap operation, and I don’t see really any use to have the Foundry object without the metadata"

```
from foundry-ml import Foundry
f = Foundry("$DOI$") # this gets the metadata, and never gets the data
f.get_data() # this downloads the data
X,y = f.load_data() #this loads the data into RAM
```

I think `get_data()` could also be called `download_dataset()`, depending on what we think is clearest. What are your thoughts?

blue442 commented:
Updated based on feedback to reflect three possible use cases:

- initialize foundry object w/o name (for publishing purposes)
- initialize foundry object w/ name, downloads metadata AND dataset
- initialize foundry object w/ name and pass `metadata_only=True` so dataset is not downloaded

Also changed function names to `download_dataset` and `download_metadata` for clarity.

Upgraded `python-setup` in github action to v4 (was causing an error).

ascourtas commented on foundry/foundry.py:
```diff
@@ -155,38 +154,24 @@ def __init__(
-                       metadata=metadata,
-                       authorizers=authorizers,
-                       interval=interval)
+        # handle empty dataset name (was returning all the datasets)
```
I'm not sure I understand this comment, could you give me a little more context?

blue442 replied:
This comment got kicked around and ended up in a weird place - I put it back in the `download_metadata` with some logic to ensure it wasn't called without a name argument, which appears to download all the datas.

ascourtas commented on foundry/models.py:
```diff
@@ -155,7 +155,7 @@ def _repr_html_(self):
     dc: Optional[Dict] = {}  # pydantic Datacite?
     mdf: Optional[Dict] = {}
-    dataset: FoundryDataset = {}
+    metadata: FoundryMetadata = {}
```
could you clarify for me the logic of removing the `dataset` field altogether?

ascourtas replied:
resolved in slack convo, removed deprecated fields

ascourtas reviewed (commented):
Just need a little clarity on things before I can approve!

blue442 commented on foundry/foundry.py:
```diff
@@ -155,38 +154,24 @@ def __init__(
+        if name:
+            self.download_metadata(name)
+
+        if name and not metadata_only:
```
Done.

blue442 commented on foundry/foundry.py:
```diff
@@ -51,7 +50,7 @@ class Foundry(FoundryBase):
     def __init__(
             self, name=None, no_browser=False, no_local_server=False, index="mdf", authorizers=None,
-            download=True, globus=True, verbose=False, metadata=None, interval=10,
+            verbose=False, metadata=None, interval=10, metadata_only=False, globus=True,
```
Things got moved around so much the docstring got lost in the shuffle - fixed!

ascourtas reviewed (changes requested):
Thanks for the changes Steve, almost there! Last thing for this PR: can you please remove the unnecessary added cells and remove the big outputs (ie for importing libraries, etc) from the data publishing notebook?

Also, where there anything you changed in the data publishing notebook that is supposed to be there? I didn't see anything but it's hard to tell without a proper diff. 

The last thing will be to update the Foundry documentation and example notebooks with the syntax change before we can cut this release.
